cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,318 @@
+"""
+This file stores a subclass of DistanceSolver, SpectralNeighborJoining. This 
+algorithm is based on the one developed by Jaffe, et al. (2021) in their paper 
+titled "Spectral neighbor joining for reconstruction of latent tree models",
+published in the SIAM Journal for Math and Data Science. 
+ 
+"""
+
+import itertools
+import networkx as nx
+import numpy as np
+import pandas as pd
+from scipy.linalg import svd
+from typing import Callable, Dict, List, Optional, Tuple
+
+from cassiopeia.solver.DistanceSolver import DistanceSolver, DistanceSolverError
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.solver import (
+    dissimilarity_functions,
+    solver_utilities,
+)
+
+
+class SpectralNeighborJoiningSolver(DistanceSolver):
+    """
+    Spectral Neighbor-Joining class for Cassiopeia.
+
+    Implements a variation on the Spectral Neighbor-Joining
+    algorithm described by Jaffe et al. in their 2020 paper. This class
+    implements the abstract class DistanceSolver. This class inherits
+    the 'solve' method from DistanceSolver with the 'find_cherry' method
+    implemented using a spectral method as in Jaffe et al.  (2020).
+
+    Args:
+        similarity_function: A function by which to compute the similarity map,
+            corresponding to the affinity R(i, j) in Jaffe et al.
+            (2020). Note that this similarity score should be multiplicative,
+            i.e. R(i, j) = R(i, k) * R(k, j) for k on the path from i to j on
+            the lineage tree. By default we will use exp(-d(i,j)) where d is the
+            metric given by weighted_hamming_distance from
+            dissimilarity_functions without weights given.
+
+        add_root: Whether or not to add an implicit root to the tree, i.e. a
+            root with unmutated characters.
+        prior_transformation: Function to use when transforming
+            priors into weights
+    """
+
+    def __init__(
+        self,
+        similarity_function: Optional[
+            Callable[
+                [np.ndarray, np.ndarray, int, Dict[int, Dict[int, float]]],
+                float,
+            ]
+        ] = dissimilarity_functions.exponential_negative_hamming_distance,
+        # type: ignore
+        add_root: bool = False,
+        prior_transformation: str = "negative_log",
+    ):
+        super().__init__(
+            dissimilarity_function=similarity_function,
+            add_root=add_root,
+            prior_transformation=prior_transformation,
+        )  # type: ignore
+
+    def get_dissimilarity_map(
+        self, cassiopeia_tree: CassiopeiaTree, layer: Optional[str] = None
+    ) -> pd.DataFrame:
+        """Outputs the first lambda matrix, where every subset is an
```

Comment:
Nit: This first line of the docs needs to be a one-liner.

Also, please describe what the "Lambda" matrix is in Jaffe et al 2020.

Also, please indicate the complexity of this procedure.

Reply:
This still is not resolved.